src/goto-programs/show_goto_functions_xml.cpp

```diff
@@ -0,0 +1,146 @@
+/*******************************************************************\
+
+Module: Goto Program
+
+Author: Thomas Kiley
+
+\*******************************************************************/
+
+#include <goto-programs/show_goto_functions_xml.h>
+#include <iostream>
+#include <sstream>
+
+
+#include <util/xml_expr.h>
+#include <util/cprover_prefix.h>
+#include <util/prefix.h>
+
+#include <langapi/language_util.h>
+
+#include "goto_functions.h"
+#include "goto_model.h"
+
+/*******************************************************************\
+
+Function: show_goto_functions_xmlt::show_goto_functions_xmlt
+
+  Inputs:
+   ns - the namespace to use to resolve names with
+
+ Outputs:
+
+ Purpose: For outputing the GOTO program in a readable xml format.
+
+\*******************************************************************/
+
+show_goto_functions_xmlt::show_goto_functions_xmlt(const namespacet &ns)
+  : ns(ns)
+{}
+
+/*******************************************************************\
+
+Function: show_goto_functions_xmlt::show_goto_functions
+
+  Inputs:
+   goto_functions - the goto functions that make up the program
+
+ Outputs:
+
+ Purpose: Walks through all of the functions in the program and returns
+          an xml object representing all their functions. Produces output
+          like this:
+          <functions>
+            <function name=main, isBodyAvaliable=true, isInternal=false>
+              <instructions>
+                <instruction id=ASSIGN>
+                  <location file="main.c" line="14"/>
+                  <instructionValue>// adsakjsdl
+                     x=y</instructionValue>
+                </instruction>
+              </instructions>
+            </function>
+          </functions>
+
+\*******************************************************************/
+
+xmlt show_goto_functions_xmlt::get_goto_functions(
```

Comment:
There is a reasonable amount of duplication involved between this and show_goto_functions_json but because it is kind of interlaced (i.e. bit of common code, bit specific, bit common) and because the types are different it wasn't clear to me exactly how I could extract the common stuff. I am open to ideas if anyone has any.